module/localisation/Localisation/src/Localisation.cpp

```diff
@@ -0,0 +1,432 @@
+            double particle_occupancy         = get_occupancy(particle_position);
+
+            // Get the position of the observation in the map for this particle [x, y]
+            Eigen::Vector2i map_position = observation_relative(particle, observation);
+
+            // Get the occupancy value of the observation at this position in the map
+            double observation_occupancy = get_occupancy(map_position);
+
+            // If the particle_occupancy is -1 then the robot is outside the map, penalise the particle
+            if (particle_occupancy == -1) {
+                weight = 0;
+            }
+            // If the observation_occupancy is -1 then the observation is outside the map, penalise the particle.
+            // Reduce the weight of the particle by the 1 / number of observations
+            else if (observation_occupancy == -1) {
+                weight -= (1.0 / n_observations);
+            }
+            else if (observation_occupancy > 0) {
+                // Increase the weight of the particle by the occupancy/n_observations, such that the maximum weight is
+                // between [0, 1]
+                weight += observation_occupancy / n_observations;
+            }
+        }
+
+        // Ensure the weight is positive
+        weight = std::max(weight, 0.0);
+
+        return weight;
+    }
+
+    Eigen::Vector3d Localisation::compute_mean() {
+        Eigen::Vector3d mean = Eigen::Vector3d::Zero();
+        for (const auto& particle : particles) {
+            mean.x() += particle.state.x();
+            mean.y() += particle.state.y();
+            mean.z() += particle.state.z();
+        }
+        return mean / cfg.n_particles;
+    }
+
+    Eigen::Matrix<double, 3, 3> Localisation::compute_covariance() {
+        Eigen::Matrix<double, 3, 3> cov_matrix = Eigen::Matrix<double, 3, 3>::Zero();
+        for (const auto& particle : particles) {
+            const Eigen::Matrix<double, 3, 1> deviation = particle.state - state;
+            cov_matrix += deviation * deviation.transpose();
+        }
+        cov_matrix /= (cfg.n_particles - 1);
+        return cov_matrix;
+    }
+
+    void Localisation::time_update() {
+        // Calculate the time since the last time update
+        using namespace std::chrono;
+        const auto current_time = NUClear::clock::now();
+        const double dt         = duration_cast<duration<double>>(current_time - last_time_update_time).count();
+        last_time_update_time   = current_time;
+
+        // Update the particles using the walk command and the time since the last time update, with some tunable
+        // scaling
+        for (size_t i = 0; i < particles.size(); i++) {
+            double delta_x         = walk_command.x() * dt * cfg.scale_x;
+            double delta_y         = walk_command.y() * dt * cfg.scale_y;
+            double delta_theta     = walk_command.z() * dt * cfg.scale_theta;
```

Comment:
hmm it'd be good to decouple the odometry walk command stuff from localisation. it isn't critical tho.